### PR TITLE
MSUnmerged parallel deletions; fetch compressed data from Rucio ConMon

### LINKED ISSUE #11792: MSUnmerged: Add multithreading for remote operations
**Impact of the new feature**
MSUnmerged

**Is your feature request related to a problem? Please describe.**
This issue is a request for feature development, following our findings we have made in the previous issue  related  to the service slowness: https://github.com/dmwm/WMCore/issues/11701#issuecomment-1791072045 

Long story short the `WebDAV` protocol is obviously slower than `SRMv2`  in communicating remote data control commands (like `listdir` and `stat`) between the server and the client. One way to mitigate this slowness is to try to avoid recursive remote operations  to the best we can. This is what we do with this PR: https://github.com/dmwm/WMCore/pull/11781 Unfortunately this won't work for every site (especially for those which are using `WebDAV` as a proxy protocol to `xrootd`, but I may be wrong here - this needs to be doublechecked).  

So the proper way to overcome the situation is to implement some sort of parallelization of remote operations, because they are implying the constraint here. And since the delays are mostly remote communication dependent it is natural to think about increasing the number of communication sessions locally - inside a single service instance (K8 pod). 

**Describe the solution you'd like**

There are two places  in `MSUnmerged`, where we can do that. The two of them reflecting a different granularity:
* We can try to spread multiple threads on a per `RSE` basis, here: https://github.com/dmwm/WMCore/blob/21342f5af95dfaa6af30aa92a714739799ed1160/src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py#L281-L284

In this approach we should create N pipelines and assign them to N threads. This way by sending one `RSE` at a time through through 1 pipelines we should scale up N times, because we will run N `RSE`s simultaneously, and expecting to reduce the I/O wait N folds. 

* We can try to spread multiple threads on a per `baseDirectory` basis, here: https://github.com/dmwm/WMCore/blob/21342f5af95dfaa6af30aa92a714739799ed1160/src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py#L325

In this approach we should create N threads in the background, and then we should simply slice this data structure: `rse['files']['toDelete']` in slices of size N  and then start feeding the N threads with the resultant N subsets of the data. This will basically create N FIFO queues, which should again supposedly reduce the I/O wait N folds.

Restrictions:
* In the first approach we will hit some memory limits, because uploading N `RSE`s in memory would be quite expensive.  
* In the second approach we already have everything in memory, so in that sense this one is much more promissing, but as will be seen down from my small scale test, this  one is actually CPU limited. 

Here is how I already implemented a Proof of concept based the second approach suggested (parallelizateion of remote operations inside the RSE): https://github.com/todor-ivanov/WMCoreAux/pull/8/commits/190eab1d647f4dd4fdafdae967414c39c7934fc9 

This is following one to one the method of splitting the data structure inside the RSE and start feeding it in N FIFO queues, as explained above. I used this code to see if we  can actually gain something and here are the results:

* Remote Operation used: `stat`
* Remote Site: `T2_IT_Legnaro`
* Hardware: Single core VM - OpenStack profile `small` 
* Number of threads: 1 vs. 10

**Single Thread:**
```
(WMCore.MSUnmergedStandalone) [@ WMCore.MSUnmergedStandalone]$ time python  $WMCORE_SERVICE_ROOT/WMCoreAux/bin/MSStandalone/MSUnmerged/init.py >  scaleTest-1_threads-T2_IT_Legnaro.log  2>&1

real	14m1.500s
user	10m15.514s
sys	0m4.998s

```

**10 Threads:**
```
(WMCore.MSUnmergedStandalone) [@ WMCore.MSUnmergedStandalone]$ time python  $WMCORE_SERVICE_ROOT/WMCoreAux/bin/MSStandalone/MSUnmerged/init.py >  scaleTest-10_threads-T2_IT_Legnaro.log  2>&1

real	8m37.584s
user	8m4.143s
sys	0m7.261s

```

* In the first case the machine was running at 70% CPU and never exceeded it 
* In the second run it was hitting the 100% constantly and all the 10 threads  were evenly sharing at around 10% CPU time each. 

The increase of CPU usage from 70% to 100% matches  almost  the gain we have in terms of runtime as well. This CPU time increase could be also due to thread managing overhead as well. This last guess of mine sounds as an overstretch, but we should not overlook this possibility as well. Even though it looks we are gaining some benefit from splitting those I/O operations in multiple threads. But for better scale test we'll need a bigger hardware. 
  
**Describe alternatives you've considered**
None 

**Additional context**
none

### LINKED ISSUE #12061: MSUnmerged T1 service crashing when pulling large storage json dump
**Impact of the bug**
MSUnmerged

**Describe the bug**
After bumping the resources limit to up to 8GB of memory RAM, ms-unmerged-t1 is still unable to load data into memory for T1_US_FNAL_Disk, which makes the service to crash every minute and repeat this in an endless mode.

In addition, several T1 storage forbid the WMCore robot certificate to delete unneeded data from the unmerged area. Identified RSEs **so far** are: CNAF and JINR.

**How to reproduce it**
Fetch the FNAL unmerged dump and load it into memory in python.

**Expected behavior**
Ideally, we should have a better control over the data that is loaded into memory, either streaming data or loading data in chunks.
We also have to follow up on those permission issues with the site admins.

**Additional context and error message**
We seem to be crashing this service at least since Mar 21st, 2024(!!!)
```
[21/Mar/2024:00:00:31]  WATCHDOG: server exited with exit code signal 9... restarting
```

### PR DESCRIPTION
**Pull request by amaltaro**
Fixes https://github.com/dmwm/WMCore/issues/12061
Fixes https://github.com/dmwm/WMCore/issues/11792

#### Status
ready

#### Description
The scope of this PR has inflated quite a lot, as I started just investigating the adoption of compressed RucioConMon data into MSUnmerged, but then had to make further modifications to actually decrease the memory footprint.

Summary of changes are:
* [RucioConMon] modified to actually stream/generate data when it's retrieved in binary mode (raw data compressed)
* [WMStatsServer] tweaked to actually return a unique list of protected LFNs
* no longer store the full list of files under `rse['dirs']['allUnmerged']`
* no longer store the full list of files under `rse['files']['allUnmerged']`
* actually, the `rse['files']` structure is not used anymore, as it can easily blow up the memory footprint
* by default, make MSUnmerged retrieve compressed data from RucioConMon (zipped option)
* removed `MSUnmerged.filterUnmergedFiles` method - logic now embedded in `getUnmergedFiles` and `_isDeletable`
* refactored `cleanRSE` method as follows:
  * first, try to delete the whole directory;
  * second, if it fails, then list all the content in that directory and try to delete only the sub-directories;
  * third, list actual files under the main directory and trigger parallel batch file deletions;
  * fourth, try again to delete those sub-directories in batches of 20 dirs per call
  * finally, try again to delete the main directory
* gfal error parsing streamlined

Memory leak (created and fixed in this PR): I have tried 10s of options to resolve the memory leak, and I managed to isolate it to the ThreadPoolExecutor objects and the gfal2 results, which don't seem to be properly freeing memory back to the application. I also found a couple of threads on this subject suggesting this leak, [here](https://github.com/python/cpython/issues/98467) and [here](https://www.adrian.idv.hk/2024-02-09-concurrent/).
After implementing the batch job as a generator, keeping executors alive as long as an RSE is being cleaned up and freeing the gfal2 context object, I can confirm that the memory utilization is finally under control and back to the usual pattern.
 
New service configuration parameters to control deletion concurrency are:
```
        self.msConfig.setdefault("parallelFileDeletionMaxWorkers", 10)
        self.msConfig.setdefault("parallelFileDeletionBatchSize", 100)
        self.msConfig.setdefault("parallelSubDirDeletionMaxWorkers", 20)  # with 1 directory per worker
```

#### Is it backward compatible (if not, which system it affects?)
YES

#### Related PRs
Gzipped support was added with this PR: https://github.com/dmwm/WMCore/pull/11142, but MSUnmerged had not been switched to use it, yet.

Standalone script to test gfal: https://gitlab.cern.ch/dmwm/wmcore-tools/-/merge_requests/14

#### External dependencies / deployment changes
Gzipped data is not currently functional:
```
>>> allUnmerged = rucioConMon.getRSEUnmerged("T1_ES_PIC_Disk", zipped=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/local/lib/python3.8/site-packages/WMCore/Services/RucioConMon/RucioConMon.py", line 127, in getRSEUnmerged
    rseUnmerged = self._getResultZipped(uri,  callname=callname, clearCache=True)
  File "/usr/local/lib/python3.8/site-packages/WMCore/Services/RucioConMon/RucioConMon.py", line 92, in _getResultZipped
    data = self._getResult(uri, callname, clearCache, args, binary=True)
  File "/usr/local/lib/python3.8/site-packages/WMCore/Services/RucioConMon/RucioConMon.py", line 73, in _getResult
    results = gzip.decompress(istream.read())
  File "/usr/local/lib/python3.8/gzip.py", line 548, in decompress
    return f.read()
  File "/usr/local/lib/python3.8/gzip.py", line 292, in read
    return self._buffer.read(size)
  File "/usr/local/lib/python3.8/gzip.py", line 479, in read
    if not self._read_gzip_header():
  File "/usr/local/lib/python3.8/gzip.py", line 427, in _read_gzip_header
    raise BadGzipFile('Not a gzipped file (%r)' % magic)
gzip.BadGzipFile: Not a gzipped file (b'/s')
```

### CODE REVIEW

**Comment by amaltaro** [comment]
I have this patch applied to the production pod, named `ms-unmer-t1-7dcf5f577b-dcj4w` and it is also using the following configuration change (inside the pod):
```
data.skipRSEs = ['T1_US_FNAL_Disk', "T1_DE_KIT_Disk", "T1_RU_JINR_Disk", "T1_UK_RAL_Disk", "T1_IT_CNAF_Disk"]
```

At least CNAF and JINR are right now failing with permission issues.

**Comment by amaltaro** [comment]
The auth/authz issues that we had before were actually coming from a missing environment variable in the manage script (X509_USER_PROXY). Fix has been provided here: https://github.com/dmwm/CMSKubernetes/pull/1532

On what concerns FNAL, I have been testing both the `json` and the `raw` format, and the memory RAM allocation gain is not very substantial (about 20%), as one can see with a simple script I wrote:
```
(WMAgent-2.3.4) [***@vocms0282:install]$ python get_unmer.py 
Size of all unmerged in unzipped format: 2489 MB and type <class 'list'>
['/store/unmerged/RunIISummer20ULPrePremix/Neutrino_E-10_gun/PREMIX/BParking_106X_upgrade2018_realistic_v16_L1v1-v1/2560000/CD81A5B4-B20C-2840-92B5-8FE53886885C.root', '/store/unmerged/RunIISummer20ULPrePremix/Neutrino_E-10_gun/PREMIX/BParking_106X_upgrade2018_realistic_v16_L1v1-v1/2560001/0DB454A7-E044-304B-AA56-2FBE12EC156E.root']

Size of all unmerged in zipped format: 1946 MB and type <class 'bytes'>
b'/s'
```

Lastly, in terms of network traffic pulling the unmerged dump for FNAL, we can see an outstanding improvement both in time and data. Here are the results:
```
(WMAgent-2.3.4) [cmst1@vocms0282:install]$ scurl "https://cmsweb.cern.ch/rucioconmon/unmerged/files?rse=T1_US_FNAL_Disk&format=json" > out_FNAL.json
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 1984M    0 1984M    0     0  71.6M      0 --:--:--  0:00:27 --:--:-- 64.3M

(WMAgent-2.3.4) [cmst1@vocms0282:install]$ scurl "https://cmsweb.cern.ch/rucioconmon/unmerged/files?rse=T1_US_FNAL_Disk&format=raw" > out_FNAL.gzip
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  225M    0  225M    0     0   162M      0 --:--:--  0:00:01 --:--:--  162M
```

**Comment by amaltaro** [comment]
After unit tests and manual checks to ensure that the final list of directories to be deleted is correct (comparing between current and new implementation), I decided to "test" it with the actual largest dump (T1_US_FNAL_Disk). See further details here:
https://github.com/dmwm/WMCore/issues/12061#issuecomment-2384621546

Functionality wise, everything seems to be working fine. Performance wise, it is still taking too long to clean up data, and the problem comes from directories with a large amount of files and/or many subdirectories. Those end up timing out (apparently 60secs) the gfal-rm operation, which then triggers a file by file deletion, e.g.:
```
2024-10-01 15:01:55,526:INFO:MSUnmerged: Processing directory index 1241 out of 13105
2024-10-01 15:01:55,526:INFO:MSUnmerged: Trying to remove the whole directory: davs://***/dcache/uscmsdisk/store/unmerged/Run3Summer23NanoAODv12/K0sTo2PiTo2Mu_K0sFilter_MuFilter_PiLifetime0p02_TuneCP5_13p6TeV_pythia8-evtgen/NANOAODSIM/130X_mcRun3_2023_realistic_v14-v2
2024-10-01 15:06:56,409:ERROR:MSUnmerged: FAILED to remove directory: davs://***/dcache/uscmsdisk/store/unmerged/Run3Summer23NanoAODv12/K0sTo2PiTo2Mu_K0sFilter_MuFilter_PiLifetime0p02_TuneCP5_13p6TeV_pythia8-evtgen/NANOAODSIM/130X_mcRun3_2023_realistic_v14-v2: gfalException: DavPosix::rmdir  Connection timed out, gfalErrorCode: 110
2024-10-01 15:06:56,410:INFO:MSUnmerged: Failed to remove the whole directory. Listing its content and deleting file by file.
2024-10-01 15:06:59,251:WARNING:MSUnmerged: Failed to list directory: davs://***/dcache/uscmsdisk/store/unmerged/Run3Summer23NanoAODv12/K0sTo2PiTo2Mu_K0sFilter_MuFilter_PiLifetime0p02_TuneCP5_13p6TeV_pythia8-evtgen/NANOAODSIM/130X_mcRun3_2023_realistic_v14-v2/2820083, gfal code: 2
2024-10-01 15:06:59,251:INFO:MSUnmerged: Entries under directory: davs://***/dcache/uscmsdisk/store/unmerged/Run3Summer23NanoAODv12/K0sTo2PiTo2Mu_K0sFilter_MuFilter_PiLifetime0p02_TuneCP5_13p6TeV_pythia8-evtgen/NANOAODSIM/130X_mcRun3_2023_realistic_v14-v2/2820083 is: 0
2024-10-01 15:07:56,248:INFO:MSUnmerged: Entries under directory: davs://***/dcache/uscmsdisk/store/unmerged/Run3Summer23NanoAODv12/K0sTo2PiTo2Mu_K0sFilter_MuFilter_PiLifetime0p02_TuneCP5_13p6TeV_pythia8-evtgen/NANOAO
DSIM/130X_mcRun3_2023_realistic_v14-v2/40003 is: 901
...
```

It could be that I removed too much code that Todor has already implemented, so I will get back to the code and make recursive deletions. Ensuring that we always try to delete a whole directory first, and only as a very last option, to go file by file.

**Comment by amaltaro** [comment]
I might have to do some polishing, but I think the bulk logic is in place now and I welcome any feedback.

File: test/python/WMCore_t/MicroService_t/MSUnmerged_t/test_gfal.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/env python
+import logging
```
Comment (amaltaro) [inline test/python/WMCore_t/MicroService_t/MSUnmerged_t/test_gfal.py:2]: SELF-REMINDER: This script will be removed before merging this PR.

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py [hunk 2]
```diff
@@ -319,120 +332,92 @@ def cleanRSE(self, rse):
...
             raise MSUnmergedPlineExit(msg) from ex
 
-        filesToDeleteCurrRSE = 0
-
         # Start cleaning one directory at a time:
-        for dirLfn, fileLfnGen in rse['files']['toDelete'].items():
-            if dirLfn in rse['dirs']['deletedSuccess']:
-                self.logger.info("RSE: %s, dir: %s already successfully deleted.", rse['name'], dirLfn)
-                continue
+        for idx, dirLfn in enumerate(rse['dirs']['toDelete']):
```
Comment (vkuznet) [inline src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:336]: As far as I can tell the `toDelete` parts of the `rse['dirs]` object is never cleaned up even though the logic below deletes dirs/files.

Comment (vkuznet) [reply src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:336]: Instead of loop I suggest to clearly define function(s) to delete an object. The function can take `fobj` (either dir or file) and perform the operation and return the status. Then code can be parallelized (to speed up nested operations) and deletion can be done concurrently.

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py [hunk 3]
```diff
@@ -306,9 +304,24 @@ def _execute(self, rseList):
     def cleanRSE(self, rse):
         """
         The method to implement the actual deletion of files for an RSE.
+        Order of deletion attempts is:
```
Comment (vkuznet) [inline src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:307]: It would be beneficial to clearly define logic of deletion, here you outline the steps instead of logic. But the underlying code seems very complex with different rules of what and how it should delete. I suggest to provide description of such logic in the doc string.

**Review by vkuznet — COMMENTED** [review]
The code seems very complex to me and very hard to follow. As such to properly define its logic someone needs to dig very deep into code flow. I suggest to further refactor the code into smaller functions with clear scope, e.g. if function is given an `fobj` which can be either a file or dir, the code can be re-organized to use concurrent patter on nested dir structure. The function behavior can be well defined with respect to its action, e.g.
- for directory get list of its content and call itself
- for a file, unlink the file and return the status

Without knowing exact logic how clean-up procedure should be done, which exceptions should be made, the review of the logic is very cumbersome.

**Comment by vkuznet** [comment]
Minor comment: the code relies on class methods/functions with underscores. I expect that those should be protected methods or not directly accessible, but python does not provide such isolation and notations neither enforce it. I don't know the rationale behind this notations.

**Comment by todor-ivanov** [comment]
one more thing @amaltaro 
about this bullet from the issue description: 
> refactored cleanRSE method by first trying to delete the whole directory. If it fails, then list all the content in that directory and delete its content by slices. Finally, try to remove the (now) empty directory. 

This  is already in place. If you are  simply changing the way how to do it with this PR, that's another story, but this is definitely nothing new for MSUnmerged

**Comment by todor-ivanov** [comment]
>    trying to remove the parent directory, if it fails
    tries to delete the subdirectories, if it fails
    get a list of contents (files) in all these subdirectories and remove by 100 file slices.
    finally, try to remove any leftover directory (which should now be empty).

> I will update the PR description with this. I also need to revisit what was already in place and see the differences.

All of this is already in place .... I do not see a point of rewriting this logic.... Well if you like it your way .. of course there is a point.

**Comment by todor-ivanov** [comment]
What needs to be done is:
* First to loosen  the initial constraints  of this service and 
* Second to isolate its (respectively ours) responsibility. 

Let me rephrase with some more details:

* We  do not need to depend on RucioConMon... neither its cycle (which basically introduces a lot of failed behavior on our side) nor the level of granularity of the scans it produces.... Since `Reqmgr` does not care about the contents  of any directory deeper than 3 levels down the tree starting from the workflow name - meaning it stops  at the highest possible level in the output - the dataset name..... and there is a solid reason for that - because from this point down the amount of information grows exponentially. Hence the disability of the service to scale ..... No matter how many tricks we apply once we  get into the spiral of recursions the complexity function is going to be nothing less but O(exp). So we do not need all this information to be transferred to us.... we  should never even try recursive deletions on a file by file basis ... and shoot ourselves in the thumb with remote recursive tree traversals (which on top of everything adds up the protocol overhead etc. etc...).  We  must stop at the level at which Reqmgr protects the data... and cut the branch there - ergo delete everything directly from this level and now down bellow. Which means we simply do not need these GBs of scanned data on a file by file basis.
* When it comes to failures to cut the branch at the level we know is protected by Reqmgr or  if we have other like cert issues, any failure gets logged  and alarmed to Site Support, and we never chaise those one by one  on  site by site basis .... we simply need to deliver previously agreed set of failures + alarms to the relevant team (if needed create  the proper REST interfaces for that - which already exist btw, and we already accumulate this information). And then the relevant team is to be assigned to communicate the resource issues with the resource providers. If we try to own all the problems of the system that would simply be a disaster if you ask me ...We won't hold such pressure.

**Comment by amaltaro** [comment]
If we don't rely on RucioConMon, how do we know which directories have been created in which storage? Are you saying that the scanning logic implemented in Rucio should become part of WMCore? I hope not!

Second, we have no choice but go down to the file level once a directory fails to be deleted (which many times it could be because it has too many content in it, from what I have seen in the logs).

About your last bullet, I totally agree that we have no way to communicate issues with the sites. However, we are responsible for the MSUnmerged functionality. Everyone has to take up responsibility in this process.

**Comment by todor-ivanov** [comment]
> If we don't rely on RucioConMon, how do we know which directories have been created in which storage? Are you saying that the scanning logic implemented in Rucio should become part of WMCore? I hope not! 

We are already doing worse .. Once we fail a deletion we  start all of this process recursively on a depth first logic && remotely..... which is a disaster. So What I suggest is to do it ourselves, but to stop at level -3. The level at we  which our services talk and never dive deeper. This is a fairly fast process and easy, because at that level of the tree the information is still in the lower (almost linear) part of the exponent.... this is something we can confidently manage ourselves.  And as I said we are already doing it .... much much much deeper..... so this needs no proof of concept.

**Comment by amaltaro** [comment]
We only scan deeper, not to the top level directories. In addition, we only scan in case of failures, because otherwise there is no way we can delete files.

Just to make sure we are speaking the same language, an example of file to be deleted would be:
```
          "/store/unmerged/RunIISummer20UL17SIM/DPSToJpsiJpsi_FourMuonFilter_SoftQCDnonD_TuneCP5_13TeV-pythia8-evtgen/GEN-SIM/106X_mc2017_realistic_v6-v3/00000/blah.root",
```

Are you saying that we need to list content up to `/store/unmerged/RunIISummer20UL17SIM/DPSToJpsiJpsi_FourMuonFilter_SoftQCDnonD_TuneCP5_13TeV-pythia8-evtgen/GEN-SIM` only? If so, you understand that to get to this point, we need to recursively list directories starting at the root directory `/store/unmerged/`, right?

I feel like we are deviating from the original goal of this issue, which is:
* to stop breaking the service and make it able to actually delete unmerged files.

We can have the redesign of MSUnmerged discussion in a different moment. For now, let us try to resolve this one issue please.

**Comment by todor-ivanov** [comment]
> Second, we have no choice but go down to the file level once a directory fails to be deleted (which many times it could be because it has too many content in it, from what I have seen in the logs).

Yes we have. It was  announced that all sites are now supposed to be using WEBDav, which btw in the past has hit us heavily, and exactly in that context I did some research and the outcome was that we  could be able to confidently and rightfully ask the recursive deletions to be supported from the sites (similarly to that we ask the write permissions at the unmerged area by certificate  role)  and we should not own all the protocol level complexities. All of which were  supposed to be encapsulated  and isolated  from us by using `gfal` - which is the whole meaning of its  sole existence btw.

**Comment by todor-ivanov** [comment]
> We can have the redesign of MSUnmerged discussion in a different moment. For now, let us try to resolve this one issue please

Your code change does not resolve an issue. With your code change you are already redesigning the service, but without having this discussion which I intentionally triggered here.

**Comment by todor-ivanov** [comment]
> I guess you didn't see the memory plots I provided in here then:
https://github.com/dmwm/WMCore/pull/12089#issuecomment-2334942594

Alan, I saw them.... I have been running similar tests  hundreds of times.... I said it few comments above. If we insist to keep depending on RucioConMon, be it, but lets separate it in a different thread.

And running any test on the current code status would be speculative.... because you did not isolate the changes in different  steps so it would be difficult to attribute the results either to some code optimizations or because you changed the way how you feed the service. So they would not be a justification  of you rewriting the rest of the  logic of the service.

**Review by todor-ivanov — APPROVED** [review]

File: src/python/WMCore/Services/RucioConMon/RucioConMon.py [hunk 4]
```diff
@@ -68,31 +67,26 @@ def _getResult(self, uri, callname="", clearCache=False, args=None, binary=False
...
-                results = gzip.decompress(istream.read())
-            return results
-        else:
-            with self.refreshCache(cachedApi, apiUrl) as istream:
-                results = istream.read()
-
-        results = json.loads(results)
-        return results
+        with self.refreshCache(cachedApi, apiUrl, decoder=True, binary=False) as istream:
+            results = istream.read()
```
Comment (vkuznet) [inline src/python/WMCore/Services/RucioConMon/RucioConMon.py:71]: NOTE: I don't know how much data istream reads, but here you have two potential memory spikes, one is reading data from istream and another to parse the json. Instead, I suggest to use
```
return json.load(istream)
```
which performs both steps as one, i.e. JSON can read from file object directly.

**Review by vkuznet — CHANGES_REQUESTED** [review]

**Comment by amaltaro** [comment]
I compared the counters for 5 RSEs, and the number of `totalNumDirs` and `dirsToDelete` match with the current code and the code provided in this PR.

I am going to run a real deletion now against FNAL_Disk, which is ever growing in the number of unmerged files (now with 13M files).

**Comment by amaltaro** [comment]
I applied this patch to the ms-unmer-t1 production pod to perform the deletion of T1_US_FNAL_Disk unneeded unmerged files - after temporarily changing the deployment limit memory RAM from 2 to 8GB.

Deletion is still ongoing and here are the expected stats:
```
2025-05-28 21:19:38,690:INFO:MSUnmerged: RSE post-filter stats for: T1_US_FNAL_Disk:     _id: 620234a9f98e512962f8cb75
    counters: 
        dirsDeletedFail: 0
        dirsDeletedSuccess: 0
        dirsToDelete: 11856
        filesDeletedFail: 0
        filesDeletedSuccess: 0
        filesToDelete: 12127786
        gfalErrors: 
        totalNumDirs: 12539
...
        ```

while Rucio ConMon stats and MSUnmerged T1 memory footprint screenshots are attached to this post.
fnal_rucio_stats
msunmerged_fnal

**Review by vkuznet — CHANGES_REQUESTED** [review]

File: test/python/WMCore_t/MicroService_t/MSUnmerged_t/test_gfal.py [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+def createGfal2Context(logLevel="normal", emulate=False):
+    """
+    Create a gfal2 context object
+    :param logLevel: string with the gfal2 log level
+    :param emulate: boolean to be used by unit tests
+    :return: the gfal2 context object
+    """
+    if emulate:
+        return None
+    ctx = gfal2.creat_context()
```
[the remark this answers is not in the document]
Comment (amaltaro) [reply test/python/WMCore_t/MicroService_t/MSUnmerged_t/test_gfal.py:20]: Believe it or not, it is the correct method name. Thank you for noticing this though.

**Review by vkuznet — CHANGES_REQUESTED** [review]

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py [hunk 6]
```diff
@@ -302,139 +308,148 @@ def _execute(self, rseList):
...
+        rse['counters'].setdefault('dirsDeletedFail', 0)
+        rse['counters'].setdefault('filesDeletedSuccess', 0)
+        rse['counters'].setdefault('filesDeletedFail', 0)
+        rse['counters'].setdefault('gfalErrors', {})
+
+        self.logger.info("Reset all counters for RSE: %s", rse['name'])
+        return rse
+
     # @profile
     def cleanRSE(self, rse):
```
Comment (vkuznet) [inline src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:336]: By looking through description and implementation I fear that `rse` object (which is nested dictionary of dirs and other attributes) may represent large memory footprint. Because it is passed to this function it already allocates and keep this memory during execution of this function. It is also unclear in docstring what is its structure and how it is scaling with growing number of directories. Few things I suggest here to provide:
- provide full schema for `rse` dictionary object and approximate memory allocation based on average RSE usage. For instance, if we have 1K directories to clean up with X LFNs it would be nice to know the size of this `rse` object.
- the implementation does provide memory printouts within individual `dirLfn` object but it does not address the amount of memory already allocated by this function call due to passed `rse` object. In other words if `rse` allocates 1GB of RAM and internally it does not increase this RAM it is fine, but it does not solve the problem of initial memory allocation passed to this function.

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py [hunk 7]
```diff
@@ -302,139 +308,148 @@ def _execute(self, rseList):
...
+        # Update the error counter
+        if errorMessage not in self.plineCounters[self.plineUnmerged.name]['gfalErrors']:
+            self.plineCounters[self.plineUnmerged.name]['gfalErrors'][errorMessage] = 0
+        self.plineCounters[self.plineUnmerged.name]['gfalErrors'][errorMessage] += 1
+
+        self.logger.error("GFAL error occurred - Code: %s, Message: %s", errorCode, errorMessage)
+
     def _rmDir(self, ctx, dirPfn):
         """
         Auxiliary method to be used for removing a single directory entry with gfal2
```
Comment (vkuznet) [inline src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:455]: Please provide full docstring descriging input parameters and output data-type

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py [hunk 8]
```diff
@@ -302,139 +308,148 @@ def _execute(self, rseList):
...
+        return rse
+
     # @profile
     def cleanRSE(self, rse):
         """
-        The method to implement the actual deletion of files for an RSE.
-        :param rse: MSUnmergedRSE object to be cleaned
-        :return:    The MSUnmergedRSE object
+        Optimized cleanup method that minimizes directory listing operations
+        and handles storage technologies that support recursive directory deletion.
```
Comment (vkuznet) [inline src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:339]: I suggest to put back description of input parameter and return object. Since `rse` is dictionary object whose size and scalability is not-known why it was a decision to return copy of this object instead of making in-place changes?

**Review by vkuznet — CHANGES_REQUESTED** [review]
It is unclear to me the usage of `rse` object through the code as it contains both input and outputs of RSE operations, including dirs, files, and counters. It seems those things should be separately tracked, one structure for input, and another structure for output (if required) and third structure for counters. The former is only required as input and potentially has a large size which scales with number of nested directories and LFNs on RSE. The other two are potentially order(s) of magnitude smaller since they only carry about failures and statistics.

I provided concrete examples in a code where knowledge of `rse` structure is crucial for determination of memory footprint. Moreover, I suggest that we should document all dictionaries structure we use in a code since it leads to lack of knowledge both for the code logic traversal as well as understanding resource requirements for code in use.

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py (hunk 8 above)
Comment (amaltaro) [reply src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:339]: I shouldn't have deleted that part of the docstring and I am adding it back. For the second comment, I will write it at the main body of the PR.

**Comment by amaltaro** [comment]
test this please

**Review by todor-ivanov — APPROVED** [review]

**Comment by amaltaro** [comment]
As this development seems to be approaching to an end, I will try to squash commits without screwing everything up 😨

**Comment by amaltaro** [comment]
@vkuznet I should have covered all the points you had in your previous review. Feel free to have another look, please.

**Review by vkuznet — COMMENTED** [review]
Alan, I raised several issues which would be good to address. But I understand that it may delay (especially as I may leave to vacation) and I have neutral opinion about the merge. Feel free to make your own judgement how to proceed.

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py [hunk 9]
```diff
@@ -302,225 +307,217 @@ def _execute(self, rseList):
...
-
-        # Finally, remove the baseDir:
-        self.logger.debug("RM baseDir: %s", baseDirPfn)
-        success = self._rmDir(ctx, baseDirPfn)
-        successList.append(success)
+            # List only the immediate sub-directories
+            for entry in ctx.listdir(dirPfn):
+                if not entry.endswith('.root'):
+                    listSubDirs.append(os.path.join(dirPfn, entry))
+        except gfal2.GError as ex:
```
[the remark this answers is not in the document]
Comment (amaltaro) [reply src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:486]: That's a very good point. Let me update this code soon.

File: src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py [hunk 10]
```diff
@@ -946,3 +907,128 @@ def getRSEList(self):
...
+        """
+        Generator that lists and yields files in batches to control memory usage.
+        :param ctx: Gfal context manager object
+        :param dirPfn: string with a directory pfn
+        :param batchSize: int with the batch size for file paths to be deleted
+        :return: generator with batches of file PFNs
+        """
+        self.logger.info("Listing files for deletion in directory: %s", dirPfn)
+
+        def processEntry(entry, currentPath):
```
[the remark this answers is not in the document]
Comment (amaltaro) [reply src/python/WMCore/MicroService/MSUnmerged/MSUnmerged.py:921]: I find it horrible too. However, at this stage, I would rather keep it here, as it has already been fully tested and I would rather avoid any regressions.

**Comment by amaltaro** [comment]
@vkuznet thank you for the review. I should have addressed all of them (and commited in the commit before the last one).

Then the last commit I was still trying to decrease the memory footprint of the service, which seems to be much better now (normal pattern?) with:
* freeing the gfal context for every directory cleaned (about 3ms of overhead per directory - so it should be not an issue even for the largest RSE I've seen so far, with ~12k directories)
* creating the thread pool (executors) at the beginning of an RSE cleanup
* cleaning/removing the thread executors at the end of an RSE cleanup

If you want to have a look only at the last 2 commits, I kept them separated for now. If you feel like you won't have time, just let me know and I think we can proceed with the merge of this PR. Thanks

**Review by vkuznet — APPROVED** [review]
Alan, thank you for addressing all issues. I have no more comments, the code looks great to me.
